jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/content/CachingHttpContentFactory.java

```diff
@@ -366,19 +365,39 @@ public String getKey()
         @Override
         public void writeTo(Content.Sink sink, long offset, long length, Callback callback)
         {
+            boolean retained = false;
             try
             {
-                _buffer.retain();
-                sink.write(true, BufferUtil.slice(_buffer.getByteBuffer(), (int)offset, (int)length), Callback.from(_buffer::release, callback));
+                retained = cachedRetain();
+                if (retained)
+                    sink.write(true, BufferUtil.slice(_buffer.getByteBuffer(), Math.toIntExact(offset), Math.toIntExact(length)), Callback.from(this::release, callback));
+                else
+                    getWrapped().writeTo(sink, offset, length, callback);
             }
             catch (Throwable x)
             {
-                // BufferUtil.slice() may fail if offset and/or length are out of bounds.
-                _buffer.release();
+                // BufferUtil.slice() may fail if offset and/or length are out of bounds,
+                // Math.toIntExact may fail too if offset or length are > Integer.MAX_VALUE.
+                if (retained)
+                    release();
                 callback.failed(x);
             }
         }
 
+        /**
+         * Atomically checks that this content still is in cache (so it hasn't been released yet and is still usable) and retain
+         * its internal buffer if it is.
+         * @return true if this content can be used and has been retained, false otherwise.
+         */
+        private boolean cachedRetain()
```

Comment:
I don't like this name, feels like "cache a retain operation"... what's wrong with just `retain()`, it is exactly what it does, and it is already on the class `CachedHttpContent`, so there is no confusion.

Reply:
How about `retainIfCached()`?

Reply:
In the backport for Jetty 12, it is just `retain()`.

`retainIfCached()` would read strange because it would always be called in an `if` statement, so `if (retainIfCached())` with the double `if`.

I guess I could be fine with the double `if`, but I really don't see why calling it differently from `retain()`.

Reply:
@sbordet what about tryRetain?

It is different from all our other reasons because it may not retain and it breaks the contract for one release for every retain call

Reply:
`tryRetain()` it is.